x-pack/plugins/files/public/components/upload_file/upload_state.ts

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  of,
+  map,
+  zip,
+  from,
+  race,
+  take,
+  Subject,
+  finalize,
+  forkJoin,
+  mergeMap,
+  switchMap,
+  catchError,
+  ReplaySubject,
+  BehaviorSubject,
+  type Observable,
+} from 'rxjs';
+import type { FileKind, FileJSON } from '../../../common/types';
+import type { FilesClient } from '../../types';
+import { i18nTexts } from './i18n_texts';
+
+import { createStateSubject, type SimpleStateSubject, parseFileName } from './util';
+
+const prop$ = <T = unknown>(initialValue: T) => new BehaviorSubject<T>(initialValue);
```

Comment:
nit: do we need this helper? I was thinking maybe explicitly writing is more readable:

```ts
public readonly uploading$ = new BehaviorSubject(false);
```

Reply:
Yeah, I was experimenting with if this might reduce some code, but I don't think it has paid off super well. I'll remove it for now.